common/src/file.rs

```diff
@@ -0,0 +1,256 @@
+    }
+}
+
+impl File for LocalFile {
+    fn read(&self) -> anyhow::Result<Vec<u8>> {
+        Ok(std::fs::read(&self.path)?)
+    }
+
+    fn read_to_string(&self) -> anyhow::Result<String> {
+        let mut file_root = std::fs::File::open(&self.path)?;
+        let mut content = String::new();
+        file_root.read_to_string(&mut content)?;
+        Ok(content)
+    }
+
+    fn read_dir(&self) -> anyhow::Result<Vec<String>> {
+        let mut files = vec![];
+        let dir_entries = fs::read_dir(&self.path)?;
+        for entry in dir_entries {
+            let entry = entry?;
+            let path = entry.path();
+            if let Some(file_name) = path.file_name() {
+                if let Some(file_name) = file_name.to_str() {
+                    files.push(file_name.to_string());
+                }
+            }
+        }
+        Ok(files)
+    }
+
+    fn create_dir_all(&self) -> anyhow::Result<()> {
+        fs::create_dir_all(&self.path)?;
+        Ok(())
+    }
+}
+
+pub struct S3File {
+    pub path: String,
+}
+
+impl S3File {
+    pub fn new(path: &str) -> Self {
+        S3File {
+            path: path.to_string(),
+        }
+    }
+}
+
+impl Write for S3File {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        let path = self.path.clone();
+        let length = buf.len();
+        let buf = buf.to_vec();
+        let handle = thread::spawn(move || {
+            let rt = Runtime::new().unwrap();
+            rt.block_on(async { s3_write_file(&path, &buf).await })
+        });
+
+        handle
+            .join()
+            .unwrap()
+            .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, format!("{}", e)))?;
+        std::result::Result::Ok(length)
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        std::result::Result::Ok(())
+    }
+}
+
+impl File for S3File {
+    fn read(&self) -> anyhow::Result<Vec<u8>> {
+        let path = self.path.clone();
+        let handle = thread::spawn(move || {
+            let rt = Runtime::new().unwrap();
+            rt.block_on(async { s3_read(&path).await })
+        });
+
+        handle.join().unwrap()
+    }
+
+    fn read_to_string(&self) -> anyhow::Result<String> {
+        let data = self.read()?;
+        Ok(String::from_utf8(data)?)
+    }
+
+    fn read_dir(&self) -> anyhow::Result<Vec<String>> {
+        let path = self.path.clone();
+        let handle = thread::spawn(move || {
+            let rt = Runtime::new().unwrap();
+            rt.block_on(async { list_files_in_s3(&path).await })
+        });
+
+        handle.join().unwrap()
+    }
+
+    fn create_dir_all(&self) -> anyhow::Result<()> {
+        let path = self.path.clone();
+        let handle = thread::spawn(move || {
+            let rt = Runtime::new().unwrap();
+            rt.block_on(async { s3_create_dir_all(&path).await })
+        });
+
+        handle.join().unwrap()
+    }
+}
+
+async fn s3_read(path: &str) -> anyhow::Result<Vec<u8>> {
```

Comment:
I think we can make those functions private at the beginning, because we want to expose XXFile, not the those functions.